### PR TITLE
Add RFC 'Generalize Probability Grid and Submap2D Structures for SDF'

### PR DESCRIPTION
**Pull request by kdaun**
[Rendered version](https://github.com/kdaun/rfcs/blob/master/text/0019-probability-grid-and-submap2d-restructuring.md)

### CODE REVIEW

File: text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Feature Name
```
Comment (gaschler) [inline text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md]: drop feature name, just put the name as a title

File: text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+Generalize Probability Grid and Submap2D Structures for Signed Distance Field Approaches
+
+## Summary
+[summary]: #summary
+
+We propose to introduce a generalized structure of the `ProbabilityGrid` and `Submap2D` to allow the utilization of alternative mapping backends such as signed distance fields.
```
Comment (gaschler) [inline text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md]: replace utilize by use. replace mapping backend by grid data structures.

File: text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+A generalized map interface allows the exploration of alternative map representations, such as Truncated Signed Distance Fields (e.g. [Kinect Fusion](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/ismar2011.pdf))  or Euclidean Signed Distance Fields (e.g. [Voxblox](https://arxiv.org/pdf/1611.03631.pdf)).
+Signed Distance Field approaches provide a straightforward extraction of iso-surfaces (eg. as surface mesh).
+First experiments indicate potential benefits in terms of accuracy of the pose, fidelity of the map and efficiency (see [2D-SDF-SLAM](https://www.researchgate.net/profile/Joscha-David_Fossel/publication/308298063_2D-SDF-SLAM_A_signed_distance_function_based_SLAM_frontend_for_laser_scanners/links/58e66d1fa6fdcc6800b47916/2D-SDF-SLAM-A-signed-distance-function-based-SLAM-frontend-for-laser-scanners.pdf)).
+
+## Approach
+[approach]: #approach
+
+We propose to introduce an abstract class `Grid2D` to serve as common interface for 2D grids.
+`Grid2D` exposes the function `GetCorrespondenceCost` which relates the correspondence of a range observation at a given position with the map.
```
Comment (gaschler) [inline text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md]: "relates" is unclear. computes a scan to grid matching difference/cost?

File: text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+`GetCorrespondenceCost` is utilized for scan matching and constraint detection instead of `GetProbability`.
```
Comment (gaschler) [inline text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md]: Is this the only function in the new interface?

File: text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+For the `ProbabilityGrid`, application of the `CorrespondenceCostFunction` yields an equivalent formulation of the underlying optimization problem as before.
+Therefore, no change in performance of the `ProbabilityGrid` is expected.
+
+Similarly, we propose to introduce an abstract class `Submap2D` serving as an interface for grid-specific implementations `Submap2DProbablityGrid` and `Submap2DTSDF`.
```
Comment (gaschler) [inline text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md]: Rename to `(Probabilistic/Occupancy)Submap2D` and `SignedDistanceSubmap2D`?

File: text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
+# Generalize Probability Grid and Submap2D Structures for Signed Distance Field Approaches
```
Comment (gaschler) [inline text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md]: in the title, you can drop SDF

File: text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+## Summary
+[summary]: #summary
+
+We propose to introduce a generalized structure of the `ProbabilityGrid` and `Submap2D` to allow the utilization of alternative grid data structures such as signed distance fields.
+`ProbabilityGrid` will be generalized to serve as an interface for scan matching and constraint detection, allowing to utilize the same scan matcher and constraint detector with different mapping backends.
```
Comment (gaschler) [inline text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md]: utilize -> use, also above

File: text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
+# Generalize Probability Grid and Submap2D Structures for Signed Distance Field Approaches
+
+## Summary
+[summary]: #summary
+
+We propose to introduce a generalized structure of the `ProbabilityGrid` and `Submap2D` to allow the utilization of alternative grid data structures such as signed distance fields.
```
Comment (gaschler) [inline text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md]: SDF is one new option, a simple one would be to add intensities to the submaps and scan matching. You could mention this, also in the motivation.

**Review by gaschler — APPROVED** [review]
Thanks, sounds good.

File: text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md (hunk 8 above)
Comment (cschuet) [reply text/0000-probability-grid-and-submap2d-restructuring-for-sdf.md]: Please do. Also I would call it "alternative grid data representations".

File: text/0000-probability-grid-and-submap2d-restructuring.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+We propose to introduce an abstract class `Grid2D` to serve as common interface for 2D grids.
+It maintains the map limits, the update indices and exposes the function `GetCorrespondenceCost` which computes a cost based on the matching of a range observation at a given position with the map.
+`GetCorrespondenceCost` is used for scan matching and constraint detection instead of `GetProbability`.
+For the `ProbabilityGrid`, application of the `CorrespondenceCostFunction` yields an equivalent formulation of the underlying optimization problem as before.
+Therefore, no change in performance of the `ProbabilityGrid` is expected.
+Furthermore, `Grid2D` exposes functions the get the minimum and maximum possible correspondence cost which are required for the constraint detection. 
+
+Similarly, we propose to rename the current `Submap2D` to `OccupancySubmap2D` and to introduce an new abstract class `Submap2D` serving as an interface for grid-specific implementations `OccupancySubmap2D` and `TruncatedSignedDistanceSubmap2D`.
+`Submap2D` exposes the grid interface `Grid2D` and the Protocol Buffers related functions `ToProto`, `UpdateFromProto` and `ToResponseProto`.
+
```
Comment (cschuet) [inline text/0000-probability-grid-and-submap2d-restructuring.md]: Can you include the class diagrams you created?

**Comment by MichaelGrupp** [comment]
Just passing by: "rendered" link doesn't work.

**Comment by kdaun** [comment]
Updated the rendered version link.

**Comment by cschuet** [comment]
@kdaun With no pending comments this seems ready to merge after a week. Can you assign RFC=0019? Thanks.

**Pull request merged**